clang/lib/Frontend/CompilerInvocation.cpp

```diff
@@ -1834,6 +1834,14 @@ bool CompilerInvocation::ParseCodeGenArgs(CodeGenOptions &Opts, ArgList &Args,
     Opts.setInlining(CodeGenOptions::NormalInlining);
   }
 
+  // If we have specified -Og and have not explicitly set -fno-extend-lifetimes,
+  // then default to -fextend-lifetimes.
+  if (Arg *A = Args.getLastArg(options::OPT_O_Group);
+      A && A->containsValue("g")) {
```

Comment:
`A->getValue() == "g"`

Perhaps move the code around ToolChains/Clang.cpp:9198 and let the driver pass -fextend-lifetimes to frontend.

Reply:
Is there an advantage to having it passed through the driver? I ask mainly because I see `-fextend-lifetimes` as being a feature of `-Og` optimization, and if you invoke `clang -cc1 -Og` directly then you'd expect/hope that the optimization behaviour would be the same as `clang -Og` - I'm not too familiar with expected Clang behaviour though, so if this is quite normal for driver/frontend behaviour then I'm fine to switch it.

Reply:
I agree, the driver's job is usually to unpack high level flags into semi-orthogonal cc1 flags. So, for example, in clang-cl, /O1 becomes a collection of things like `-Os -ffunction-sections -fdata-sections` and maybe something else I've forgotten. `clang++ -fexceptions` becomes `-fexceptions -fcxx-exceptions` in most contexts, separating destructor cleanups from try/throw language support.

cc1 flags kind of represent features that are useful to toggle on and off for testing and test case reduction, since they're one of the things we reduce in the [clang crash reducer script](https://github.com/llvm/llvm-project/blob/main/clang/utils/creduce-clang-crash.py).

Reply:
I've moved this behaviour to the driver now. Also regarding the first request to use `A->getValue() == "g"` - if just `-O` is passed then `A->getValue()` will crash, as there are no values. An alternative would be to change the check to `A->getNumValues() && A->getValue() == "g"`, but I think `containsValue` is more concise.